doc/config.js.md
>**[Return to the Masterportal documentation](doc.md)**

[TOC]

# config.js

The `config.js` contains Masterportal configuration not directly related to UI or layers. For example, paths to other configuration files belong here. This file is usually placed next to the `index.html` and `config.json` files.

In the following, all configuration options are described. For all configuration options of type `object`, further nested options are linked and described in detail after the main table. You may also refer to **[this config.js example file](https://bitbucket.org/geowerkstatt-hamburg/masterportal/src/dev/portal/basic/config.js)**.

|Name|Required|Type|Default|Description|Example|
|----|--------|----|-------|-----------|-------|
|layerConf|yes|String||Path to the **[services.json](services.json.md)** file containing all available WMS layers and WFS feature types. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/services-internet.json"`|
|namedProjections|yes|String[]||Definition of the usable coordinate systems. See **[syntax definition](http://proj4js.org/#named-projections)** for details..|`[["EPSG:25832", "+title=ETRS89/UTM 32N +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"]]`|
|restConf|yes|String||Path to the **[rest-services.json](rest-services.json.md)** file describing further services, e.g. print service, WPS, CSW. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/rest-services-internet.json"`|
|styleConf|yes|String||Path to the **[style.json](style.json.md)** file describing vector layer (WFS) styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/style.json"`|
|addons|no|String[]|`[]`|List of names for custom modules. The modules are to be placed in the folder `/addons/`, with their entry points being defined in the `addonsConf.json`.|`["myAddon1", "myAddon2"]`|
|alerting|no|**[alerting](#markdown-header-alerting)**|`{"category": "alert-info", "isDismissable": true, "isConfirmable": false, "position": "top-center", "fadeOut": null}`|Overrides the alert module's default values.|{fadeOut: 6000}|
|cameraParameter|no|**[cameraParameter](#markdown-header-cameraparameter)**||Initial camera parameter||
|cesiumParameter|no|**[cesiumParameter](#markdown-header-cesiumparameter)**||Cesium flags||
|clickCounter|no|**[clickCounter](#markdown-header-clickcounter)**||Configuration of the clickCounter module that loads an iFrame per click.||
|cswId|no|String|`"3"`|Reference to a CSW interface used to retrieve layer information. The ID will be resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`"my CSW-ID"`|
|defaultToolId|no|String|`"gfi"`|The tool with the given ID will be active when no other tool is active.|"filter"|
|featureViaURL|no|**[featureViaURL](#markdown-header-featureviaurl)**||Optional configuration for the URL parameter `featureViaURL`. See **[urlParameter](urlParameter.md)** for details. Implemented for treeTypes *light* and *custom*.||
|footer|no|**[footer](#markdown-header-footer)**||If set, a footer is shown and configured with this object.||
|gfiWindow|no|String|`"detached"`|_Deprecated in the next major release. Please use the attribute "Portalconfig.menu.tool.gfi.desktopType" of the **[config.json](#config.json.md)** instead._ Display type and attribute information for all layer types. **attached**: the attribute information window is opened at click position **detached**: the attribute information window is opened at the top right of the map; a marker is set to the click position.|`"attached"`|
|ignoredKeys|no|String[]|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|List of attribute names to be ignored for attribute information lists of all layer types. Only used with "gfiAttributes": "showAll".|`["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH","GEOM"]`|
|infoJson|no|String|`"info.json"`|Path to the `info.json` file containing additional information on snippets. The path is relative to the *index.html*.|`"info.json"`|
|inputMap|no|Object|`{}`|If this object is set, and its field `setMarker` is set to true, the Masterportal is configured as input element for data. In that case, each click sets a Map Marker and communicates the coordinates via **[remoteInterface](remoteInterface.md)** in the chosen coordinate reference system.|`{setMarker: true, targetProjection: "EPSG:4326", setCenter: false}`|
|inputMap.setCenter|no|Boolean|`false`|Center on a marker after producing it?|`setCenter: true`|
|inputMap.setMarker|no|Boolean|`false`|Flag to activate the 'setMarker' functionality.|`setMarker: true`|
|inputMap.targetProjection|no|String|`"EPSG:25832"`|The target coordinate reference system. Coordinates will be translated to it before being communicated via **[remoteInterface](remoteInterface.md)**.|`targetprojection: "EPSG:4326"`|
|mapMarker|no|**[mapMarker](#markdown-header-mapmarker)**||Overrides the map marker module's default values. Useful for 3D markers since OpenLayers's overlays can not be displayed in 3D mode. For this, the map marker has to be defined as vector layer.||
|metaDataCatalogueId|no|String|`"2"`|URL to the metadata catalog linked to in the layer information window. The ID is resolved to a service of the **[rest-services.json](rest-services.json.md)** file.|`"MetaDataCatalogueUrl"`|
|metadata|no|**[metadata](#markdown-header-metadata)**||Allows configuration of which metadata URLs are to be resolved via proxy.||
|mouseHover|no|**[mouseHover](#markdown-header-mousehover)**||Activates the MouseHover feature for vector layers, both WFS and GeoJSON. For per-layer configuration, see the **[config.json](config.json.md)**'s section *Themenconfig.Fachdaten.Layer*.|`true`|
|obliqueMap|no|Boolean|`false`|If set to `true`, an oblique map layer is created. An additional oblique layer must be defined.||
|portalConf|no|String|`"config.json"`|Path to the portal's `config.json` file. You may also enter a node; in that case the taken path is controlled by the urlParameter `config`.|Direct path: "../masterTree/config.json"; Node: "../../portal/master/". In the node scenario, a query parameter like `config=config.json` must exist in the URL.|
|postMessageUrl|no|String|`"http://localhost:8080"`|URL the portal is supposed to post messages to and receive messages from with the `postMessage` feature.|"http://localhost:8080"|
|proxyHost|no|String||Host name of a remote proxy with CORS configured to support the portal's domain, among others.|`"https://proxy.example.com"`|
|quickHelp|no|Object|`{}`|Activates the QuickHelp module. This displays a window containing help text for supported functions of the modules. Available for the layer tree (CustomTree), the search bar (Searchbar), and the measuring tool (MeasureTool).||
|remoteInterface|no|**[remoteInterface](#markdown-header-remoteinterface)**||Optional remote interface configuration.||
|scaleLine|no|Boolean|`false`|Controls whether a scale line is displayed at the bottom right of the map.|`true`|
|simpleMap|no|Boolean|`false`|_Deprecated in the next major release. Please use the parameter `simpleMap` as part of the configuration of the `saveSelection` tool in the **[config.json](config.json.md)**._ Adds a SimpleMap URL to the `Save selection` dialogue. When calling this URL, the menu bar, layer tree, and map controls are deactivated. Not implemented for tree type *„light“*.|`false`|
|startingMap3D|bi|Boolean|`false`|Controls whether the map should start in 3D mode.||
|tree|no|**[tree](#tree)**||||
|uiStyle|no|String|`"default"`|Sets the control element layout. |`table`|
|wfsImgPath|no|String||Path to the folder holding images for the WFS styles. The path is relative to *js/main.js*.|`https://geodienste.hamburg.de/lgv-config/img/"`|
|wpsID|no|String|`""`|Reference to a WPS interface used in various modules. The ID is resolved to a service defined in the **[rest-services.json](rest-services.json.md)** file.|`""`|
|zoomToFeature|no|**[zoomToFeature](#markdown-header-zoomtofeature)**||Optional configuration of the URL query parameter `featureid`. For details, see **[urlParameter](urlParameter.md)**.||

***

## alerting

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|fetchBroadcastUrl|no|String|`false`|The alerting module will initially use a linked configuration file from this URL, if set.|
|localStorageDisplayedAlertsKey|no|String|`"displayedAlerts"`|Arbitrary key used to store information regarding the alerting module in the browser's local storage.|

***

## cameraParameter

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|heading|no|Number||Camera's initial heading in radians|
|tilt|no|Number||Camera's initial tile in radians|
|altitude|no|Number||Camera's initial height in meters|

***

## cesiumParameter
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|fog|no|Object||Fog options. See [fog documentation](https://cesiumjs.org/Cesium/Build/Documentation/Fog.html) for details.|
|enableLighting|no|Boolean|`false`|Activates light effects on the map based on the sun's position.|
|maximumScreenSpaceError|no|Number|`2.0`|Detail level in which terrain/raster tiles are fetched. 4/3 is the highest quality level.|
|fxaa|no|Boolean|`true`|activates *fast approximate anti-aliasing*|
|tileCacheSize|no|Number|`100`|terrain/raster tile cache size|

***

## clickCounter
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|desktop|no|String||iFrame URL for desktop applications|
|mobile|no|String||iFrame URL for mobile applications|

**Example:**

```json
{
    "clickCounter": {
        "desktop": "http://static.hamburg.de/countframes/verkehrskarte_count.html",
        "mobile": "http://static.hamburg.de/countframes/verkehrskarte-mobil_count.html"
    }
}
```

***

## footer

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|urls|no|**[urls](#markdown-header-footerurls)**[]||Array of URL configuration objects.|
|showVersion|no|Boolean|`false`|If `true`, the Masterportal version number is included in the footer.|

***

### footer.urls
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|alias|no|String|`"Landesbetrieb Geoinformation und Vermessung"`|Link text for desktop applications|
|alias_mobil|no|String|`"LGV"`|Link text for mobile applications|
|bezeichnung|no|String|`"Kartographie und Gestaltung: "`|Link prefix|
|url|no|String||the URL the `alias` (or `alias_mobile`) text links to, e.g. `"https://example.com/"`. You may also link to mail creation with e.g. `"mailto:my_mail@example.com"`|
|toolModelId|no|String|`"sdpdownload"`|The id of a module to be opened on clicking the link. Do not include a URL for such cases.|

**Example:**

```json
{
    "footer": {
        "urls": [
            {
                "bezeichnung": "Cartography and design: ",
                "url": "https://geoinfo.hamburg.de/",
                "alias": "Landesbetrieb Geoniformation und Vermessung",
                "alias_mobil": "LGV"
            },
            {
                "bezeichnung": "",
                "url": "http://www.hamburg.de/bsu/timonline",
                "alias": "Map inconsistencies"
            },
            {
                "bezeichnung": "",
                "url": "",
                "alias": "SDP Download",
                "toolModelId": "sdpdownload"
            }
        ],
        "showVersion": true
    }
}
```

***

## mapMarker

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|pointStyleId|no|String|`"defaultMapMarkerPoint"`|StyleId to refer to a `style.json` point style. If not set, the `img/mapMarker.svg` is used.|
|polygonStyleId|no|String|`"defaultMapMarkerPolygon"`|StyleId to refer to a `style.json` polygon style.|

**Example:**

```json
{
    "mapMarker": {
        "pointStyleId": "customMapMarkerPoint",
        "polygonStyleId": "customMapMarkerPolygon"
    }
}
```

***

## mouseHover

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|minShift|no|Integer|`5`|Minimum mouse position movement required to render a new tooltip; in pixels.|
|numFeaturesToShow|no|Integer|`2`|Maximum amount of element information per tooltip; when exceeded, an information text informs the user of cut content.|
|infoText|no|String|`"(Further objects. Please zoom.)"`|Information text shown when `numFeaturesToShow` is exceeded.|

***

## portalLanguage

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|enabled|yes|Boolean|`true`|Controls whether a button to switch the portal's language is provided.|
|debug|no|Boolean|`false`|Controls whether debug information regarding translations is logged to the console.|
|languages|yes|Object|`{ de: "deutsch", en: "englisch" }`|Language abbreviations. Please mind that matching locale files must exist.|
|fallbackLanguage|no|String|`"de"`|Fallback language used if contents are not available in the currently selected language.|
|changeLanguageOnStartWhen|no|String[]|`["querystring", "localStorage", "navigator", "htmlTag"]`|Order of user language detection. See [i18next browser language detection documentation](https://github.com/i18next/i18next-browser-languageDetector) for details.|
|loadPath|no|String|`"/locales/{{lng}}/{{ns}}.json"`|Path to load language files from, or a function returning such a path: `function(lngs, namespaces) { return path; }`. `lng` and `ns` are read from the path, if given, as if from a static path. You may also provide a URL like `"https://localhost:9001/locales/{{lng}}/{{ns}}.json"`. See [i18next http backend documentation](https://github.com/i18next/i18next-http-backend) for details.|

**Example:**

```json
{
"portalLanguage": {
        "enabled": true,
        "debug": false,
        "languages": {
            "de": "deutsch",
            "en": "englisch"
        },
        "fallbackLanguage": "de",
        "changeLanguageOnStartWhen": ["querystring", "localStorage", "navigator", "htmlTag"],
        "loadPath": "/locales/{{lng}}/{{ns}}.json"
    }
}
```

***

## quickHelp

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|imgPath|no|String|`"/"`|An absolute or relative path to the folder containing the quick help images.|
|searchbarAllgemeines1|no|String|`"allgemein.png"`|First quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarAllgemeines2|no|String|`"allgemein_2.png"`|Second quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarAllgemeines3|no|String|`"allgemein_3.png"`|Third quick help image regarding the Searchbar, belonging to the chapter "General information". The image must exist in the `imgPath` folder.|
|searchbarFlurstueckssuche|no|String|`"allgemein_4.png"`|Quick help image regarding the Searchbar, belonging to the chapter "Parcel search". The image must exist in the `imgPath` folder.|
|aufbau1|no|String|`"themen.png"`|First quick help image regarding the layer tree (CustomTree), belonging to the structure chapter. The image must exist in the `imgPath` folder.|
|aufbau2|no|String|`"themen_2.png"`|Second quick help image regarding the layer tree (CustomTree), belonging to the structure chapter. The image must exist in the `imgPath` folder.|
|statistikFlaecheNiemeier|no|String|`"Statistik_Flaeche_Niemeier.png"`|First quick help image regarding the measuring tool (MeasureTool), belonging to the "Statistical approximation" chapter. The image must exist in the `imgPath` folder.|
|statistikStreckeUniErlangen|no|String|`"Statistik_Strecke_UniErlangen.png"`|Second quick help image regarding the measuring tool (MeasureTool), belonging to the "Statistical approximation" chapter. The image must exist in the `imgPath` folder.|
|utmStreifen|no|String|`"UTM_Streifen.png"`|First quick help image regarding the measuring tool (MeasureTool), belonging to the "Equalization in UTM" chapter. The image must exist in the `imgPath` folder.|
|utmVerzerrung|no|String|`"UTM_Verzerrung.png"`|Second quick help image regarding the measuring tool (MeasureTool), belonging to the "Equalization in UTM" chapter. The image must exist in the `imgPath` folder.|
|utmFormeln|no|String|`"UTM_Formeln.png"`|Third quick help image regarding the measuring tool (MeasureTool), belonging to the "Equalization in UTM" chapter. The image must exist in the `imgPath` folder.|

***

## remoteInterface
|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|postMessageUrl|no|String|`"http://localhost:8080"`|URL the portal will post to and receive messages from with the `postMessage` feature.|

**Example:**

```json
{
    "remoteInterface": {
        "postMessageUrl": "http://localhost:8080"
    }
}
```
***

## tree

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|orderBy|no|String|`"OpenData"`|Category the layer tree is sorted by initially.|
|layerIDsToIgnore|no|Array||Array of `services.json` layer ids not to be shown in the layer tree.|
|layerIDsToStyle|no|**[layerIDsToStyle](#markdown-header-treelayeridstostyle)**[]||Special implementation for a HVV (Hamburg public transportation) service. Contains objects to request various styles of a layer id.|
|metaIDsToMerge|no|String[]||All layers found in the `services.json` regarding these meta IDs are merged to a single layer of the layer tree.|
|metaIDsToIgnore|no|String[]||All `services.json` layers listed will not be shown in the layer tree.|
|isFolderSelectable|no|Boolean|`true`|Globally sets whether a selection box is provided on folders that de-/activates all layers in it. An override per element exists, see **[config.json](config.json.md#Ordnerkonfiguration-Fachdaten)**.|

***

### tree.layerIDsToStyle

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|id|no|Sring||a `services.json` layer's id|
|styles|no|String/String[]||Style to be used as string; if multiple styles are to be used, they are listed in an array.|
|name|no|String/String[]||Name to be used as string; if multiple names are to be used, they are listed in an array.|
|legendUrl|no|String/String[]||Legend image URL to be used as string; if multiple legend images are to be used, their URLs are listed in an array.|

**Example:**

```json
{
    "tree": {
        "orderBy": "opendata",
        "layerIDsToIgnore": ["1912", "1913"],
        "layerIDsToStyle": [
            {
                "id": "1935",
                "styles": ["geofox_Faehre", "geofox-bahn", "geofox-bus", "geofox_BusName"],
                "name": ["Fährverbindungen", "Bahnlinien", "Buslinien", "Busliniennummern"],
                "legendURL": ["http://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-faehre.png", "http://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-bahn.png", "http://geoportal.metropolregion.hamburg.de/legende_mrh/hvv-bus.png", "http://87.106.16.168/legende_mrh/hvv-bus.png"]
            }
        ],
        "metaIDsToMerge": [
            "FE4DAF57-2AF6-434D-85E3-220A20B8C0F1"
        ],
        "metaIDsToIgnore": [
            "09DE39AB-A965-45F4-B8F9-0C339A45B154"
        ],
        "isFolderSelectable": false
    }
}
```

***

## metadata

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|useProxy|no|String[]||_Deprecated in the next major release. *[GDI-DE](https://www.gdi-de.org/en)* recommends setting CORS headers on the required services instead._ Describes which metadata URLs are to be requested via proxy. The request will contain the requested URL as path, with dots replaced by underscores.|

**Example:**

```json
{
    "metadata": {
        "useProxy": [
            "https://metaver.de/csw"
        ]
    }
}
```

***

## zoomToFeature

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|imgLink|yes|String||Marker link. _Deprecated in the next major release._|
|wfsId|yes|String||ID to a WFS layer of which features to a position are requested from.|
|attribute|yes|String||Attribute by which the WFS is filtered.|
|styleId|no|String||A styleId from the `styles.json` may be supplied to override the map marker's design|
|useProxy|no|Boolean|`false`|_Deprecated in the next major release. *[GDI-DE](https://www.gdi-de.org/en)* recommends setting CORS headers on the required services instead._ Whether the service URL is to be requested via proxy. The request will contain the requested URL as path, with dots replaced by underdashes.|

**Example:**

```json
{
    "zoomtofeature": {
        "attribute": "flaechenid",
        "wfsId": "4560",
        "styleId": "location_eventlotse"
    }
}
```

***

## zoomToGeometry

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|imgLink|yes|String||Marker link. _Deprecated in the next major release._|
|layerId|yes|String|`"123456789"`|Id of the WFS layer the geometry is requested from.|
|attribute|yes|String|`"district_name"`|Attribute by which the WFS is filtered.|
|geometries|yes|String|`["DISTRICT1", "DISTRICT2"]`|Contains the geometries to be filtered from the WFS.|

**Example:**

```json
{
    "zoomToGeometry": {
        "layerId": "123456789",
        "attribute": "district_name",
        "geometries": ["DISTRICT1", "DISTRICT2"]
    }
}
```

***

## featureViaURL

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|epsg|no|Integer|`4326`|EPSG code for coordinate reference system to translate coordinates to.|
|layers|yes|**[layers](#markdown-header-featureviaurllayers)**[]||Layer configuration array for given features.|
|zoomTo||String/String[]||Id of **[layers](#markdown-header-featureviaurllayers)** or array thereof, to which the Masterportal initially zooms. If none are given, the usual initial center coordinate is used.|

**Example:**

```json
{
    "featureViaURL": {
        "epsg": 25832,
        "zoomTo": "urlPointFeatures",
        "layers": [
            {
                "id": "urlPointFeatures",
                "geometryType": "Point",
                "name": "URL Point Features",
                "styleId": "url_points"
            },
            {
                "id": "urlLineFeatures",
                "geometryType": "LineString",
                "name": "URL Line Features",
                "styleId": "url_lines"
            },
            {
                "id": "urlPolygonFeatures",
                "geometryType": "Polygon",
                "name": "URL Polygon Features",
                "styleId": "url_polygons"
            },
            {
                "id": "urlMultiPointFeatures",
                "geometryType": "MultiPoint",
                "name": "URL MultiPoint Features",
                "styleId": "url_mulitpoints"
            },
            {
                "id": "urlMultiLineStringFeatures",
                "geometryType": "MultiLineString",
                "name": "URL MultiLineString Features",
                "styleId": "url_multilinestring"
            },
            {
                "id": "urlMultiPolygonFeatures",
                "geometryType": "MultiPolygon",
                "name": "URL MultiPolygon Features",
                "styleId": "url_multipolygons"
            }
        ]
    }
}
```

***

### featureViaURL.layers

The parameters described apply for each entry of the **[layers](#markdown-header-featureviaurllayers)** array.

|Name|Required|Type|Default|Description|
|----|--------|----|-------|-----------|
|id|yes|String||unique ID for the layer to be created|
|geometryType|yes|enum["LineString", "Point", "Polygon", "MultiPoint", "MultiLineString", "MultiPolygon"]||Geometry type of the feature to be shown.|
|name|yes|String||Layer name displayed in the layer tree, the legend, and the GFI pop-up.|
|styleId|no|String||Style id to be used for the feature, referring to the **[style.json](style.json.md)**.|

**Example:**

```json
{
    "layers": [{
        "id": "urlPolygonFeatures",
        "geometryType": "Polygon",
        "name": "URL Polygon Features",
        "styleId": "url_polygons"
    }]
}
```

***

>**[Masterportal translation documentation](languages.md)**

>**[Return to the Masterportal documentation](doc.md)**


CHANGELOG.md
# Changelog Masterportal
 All important changes in this project are stored in this file.

[Semantic versioning](https://semver.org/spec/v2.0.0.html) is used.

## Known Issues
- 3D: The position indicator inside of a 3D object vanishes when clicking on the object.

---

## Unreleased - in development
### Added
- New attribute 'nearbyTitle' implemented in config.json for the title in the list of nearby search results.
- Add @babel/eslint-parser to the package.json

### Changed
- Renamed the folders `library` and `util` -> `utils`, `test` -> `tests` and `ressources` -> `resources`.
- Tool addons are now also written in config.json in camelCase.
- colorTools are renamed into convertColor (src/utils/convertColor)
- Updates the core-js and babel dependencies in the package.json
- Update the dependency caniuse-lite.
- The module addGeoJSON switched from backbone to vue and is provided as a util now.
- The loading image is now displayed longer when switching to map mode: Oblique (max. 80000ms).

### Deprecated
- colorArrayToRgb (src/utils/colorArrayToRgb) is deprecated. Use convertColor (src/utils/convertColor) instead.

### Removed
- remove babel-eslint from the package.json
- Documentation files in German language have been removed. From now on the documentation will only be maintained in English. The only exception is config.json.de.md, because a German language file is necessary for the mp-admin.

### Fixed
- Fixed a bug that prevented the gfi of a wms from being requested via a reverseproxy.
- Printing wfs cluster features and features with multiple styles.
- If the transparency of a layer is defined as string, it will be parsed to an integer now.
- When clicking on layers of type TERRAIN3D, TILESET3D or OBLIQUE in the search, a query is now displayed that can be used to switch to the respective map mode.
- Fixed a bug that for the gfi in infoFormat "text/html" did not execute the script part when starting the iFrame.
- Fixed an issue in searchbar, that after finding a result and zooming to it the 3D mode couldn't be activated anymore
- Fixed an issue in fileImport that when the imported KML File is missing the isVisible setting the feature wouldn't be displayed when printing
---

## v2.8.0 - 2021-04-07
### Added
- External WMS Layers with version lower than 1.3.0 can also be imported.
- Added possibility to prepend prefix to GFI attribute.
- WMTS Layers can now be printed with MapFish.
- The user can search visible wfs features by clicking a position on the map by clicking the "Nearby button" on the controls
- The user can search visible wfs features after searching an address from search field by clicking the "Nearby button" on the controls

### Changed
- The tool AddWMS switched from backbone to vue module.
- Consolidate the mqttOptions in the layer sensor with the documentation (services.json.md).
- In config.json.md, the attribute `isActive` was changed to `active` in Portalconfig.menu.tool.gfi.
- Legends are only printed from layers that are switched visible.
- Layers of type `SensorThings` are now automatically displayed in the default topic tree. The attribute `related_wms_layers` can be used to hide related layers of type `WMS`.

### Removed
- SensorThings address tab in layerInformation removed.
- The print module for using mapfishprint v2 has been removed. In this case the attribute `proxyurl` has been removed too
- The vector style module for using the style_v2.json has been removed. In this case the atteibute `useVectorStyleBeta` has been removed too.

### Fixed
- GFI of type text/html now also loads css files of the body.
- Show metadata if the MD_Identification node is implemented as MD_DataIdentification or SV_ServiceIdentification.
- Fixed a bug that made single layer info legend appear multiple times when activating additional layers while opened.
- Printing of tooltip of measure-tool works: The tooltip of the measure tool is no longer implemented as an overlay, it is created now as vectorlayer.
- Printing of big amount of features now works.
- Removed a bug causing an irritating map focus after using BKG search and leave search field

---

## v2.7.2 - 2021-03-09
### Fixed
- SensorThings: Fix the bug where initially data has not been fetched, resulting display of "no data" on mousehover.

---

## v2.7.1 - 2021-03-05
### Fixed
- The representation in the legend has been corrected.

---

## v2.7.0 - 2021-03-03
### Added
- Integration and use of WMS-services secured via User / Password (HTTP-Basic). Thy are marked in the topic tree by lock symbol.
- On draw tool Double Circle configuration available in config.json
- On draw tool Tooltip and Tooltip style configuration available in config.json
- On tools Imported Layer in theme tree integrated function under Addons
- On tools Geometrie/Gebäude Analyse funtion on tools under Addons
- On tools Taktische Zeichen on tools under Addons
- Added the parameters 'contactInfo', 'includeSystemInfo', 'locationOfCustomerService', 'maxLines', 'showPrivacyPolicy' and 'subject' to the configuration of the contact tool.

### Changed
- The language of the changelog has been changed to English.
- Migrated the SearchByCoord Tool from Backbone.js to Vue.js.
- The measure tool has been migrated to Vue.
- Migrated the Download Tool from Backbone.js to Vue.js and integrated it into the Draw Tool.
- Migrated the SaveSelection Tool from Backbone.js to Vue.js. It also no longer has a Radio channel.
- copyToClipboard is now a reusable action and can no longer be triggered through the Radio.
- Migrated the Contact Tool from Backbone.js to Vue.js. In doing so, the functionality of the config parameter 'deleteAfterSend' was split into two parameters; 'closeAfterSend' and 'deleteAfterSend'.
- Migrated the StyleVT Tool from Backbone.js to Vue.js.
- WPS migrated to Vue.js.

### Deprecated
- Deprecated the parameter 'serviceID' for the contact tool; 'serviceId' should be used instead.

portal/master/config.js
const Config = {
    addons: ["populationRequest", "vueAddon", "tacticalMark", "trinkwasser", "schulinfo", "continuousCountingBike", "verkehrsstaerken", "solaratlas", "dataTable", "reisezeiten"],
    alerting: {
        fetchBroadcastUrl: "https://geodienste.hamburg.de/lgv-config/newsFeedPortalAlerts.json"
    },
    ignoredKeys: ["BOUNDEDBY", "SHAPE", "SHAPE_LENGTH", "SHAPE_AREA", "OBJECTID", "GLOBALID", "GEOMETRY", "SHP", "SHP_AREA", "SHP_LENGTH", "GEOM"],
    wfsImgPath: "https://geodienste.hamburg.de/lgv-config/img/",
    zoomToFeature: {
        attribute: "flaechenid",
        wfsId: "4560",
        styleId: "location_eventlotse"
    },
    metadata: {
        useProxy: [
            "https://metaver.de/csw"
        ]
    },
    zoomToGeometry: {
        layerId: "1692",
        attribute: "bezirk_name",
        geometries: [
            "ALTONA",
            "HARBURG",
            "HAMBURG-NORD",
            "BERGEDORF",
            "EIMSBÜTTEL",
            "HAMBURG-MITTE",
            "WANDSBEK"
        ]
    },
    namedProjections: [
        // GK DHDN
        ["EPSG:31467", "+title=Bessel/Gauß-Krüger 3 +proj=tmerc +lat_0=0 +lon_0=9 +k=1 +x_0=3500000 +y_0=0 +ellps=bessel +datum=potsdam +units=m +no_defs"],
        // ETRS89 UTM
        ["EPSG:25832", "+title=ETRS89/UTM 32N +proj=utm +zone=32 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs"],
        // LS 320: zusammen mit Detlef Koch eingepflegt und geprüft
        ["EPSG:8395", "+title=ETRS89/Gauß-Krüger 3 +proj=tmerc +lat_0=0 +lon_0=9 +k=1 +x_0=3500000 +y_0=0 +ellps=GRS80 +datum=GRS80 +units=m +no_defs"],
        // WGS84
        ["EPSG:4326", "+title=WGS 84 (long/lat) +proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs"]
    ],
    footer: {
        urls: [{
            "bezeichnung": "common:modules.footer.designation",
            "url": "https://geoinfo.hamburg.de/",
            "alias": "Landesbetrieb Geoinformation und Vermessung",
            "alias_mobil": "LGV"
        }],
        showVersion: true
    },
    quickHelp: {
        imgPath: "https://geodienste.hamburg.de/lgv-config/img/"
    },
    cswId: "3",
    metaDataCatalogueId: "2",
    portalConf: "./",
    layerConf: "https://geodienste.hamburg.de/services-internet.json",
    restConf: "https://geodienste.hamburg.de/lgv-config/rest-services-internet.json",
    styleConf: "https://geodienste.hamburg.de/lgv-config/style_v3.json",
    scaleLine: true,
    mouseHover: {
        numFeaturesToShow: 2,
        infoText: "(weitere Objekte. Bitte zoomen.)"
    },
    clickCounter: {},
    startingMap3D: false,
    obliqueMap: true,
    cesiumParameter: {
        tileCacheSize: 20,
        enableLighting: true,
        fog: {
            enabled: true,
            density: 0.0002,
            screenSpaceErrorFactor: 2.0
        },
        maximumScreenSpaceError: 2,
        fxaa: true
    },
    defaultToolId: "gfi",
    portalLanguage: {
        enabled: true,
        debug: false,
        languages: {
            de: "deutsch",
            en: "englisch",
            it: "italienisch"
        },
        fallbackLanguage: "de",
        changeLanguageOnStartWhen: ["querystring", "localStorage", "navigator", "htmlTag"]
    }
};

// conditional export to make config readable by e2e tests
if (typeof module !== "undefined") {
    module.exports = Config;
}
